play loop + refactoring

TGLuis opened the pull request:
solve #48

TGLuis committed:
play loop + refactoring

FrancoisDuchene reviewed (approved):
monsieur propre monsieur propre 🎵🧽

Jimvy commented:
Changer de fragment interrompt le son, c'est voulu ?

Jimvy commented:
Ah sorry, ouvrir les paramètres ça coupe le son

TGLuis commented:
J'ai pas ce soucis sur mon gsm :/ maybe Android 10? Tu as sû identifier ce qui faisait ça dans le code/les logs?

Jimvy commented:
Ok, donc dès qu'on arrive sur le fragment "Paramètres" (via le drawer ou via la roue dentée), ça coupe le son, mais l'animation du bouton continue à jouer...
J'ai pas encore regardé dans les logs. Je confirme que c'est présent sur mon téléphone et sur l'émulateur, tous deux sur Android 10.
Ca commencerait quand même à faire beaucoup de "bugs" introduits par Android 10 non ?

TGLuis commented:
T'avais pas le soucis avant ce commit?

Jimvy commented:
Aussi, quand je downgrade vers une version antérieure de l'app, ça crash... parce qu'il connait évidemment pas "LOOP" comme mode de lecture

Jimvy commented:
Bien, j'ai été regarder dans les logs : en gros, dès qu'on ouvre le fragment des paramètres, le listener du spinner est appelé 2 fois et set le paramètre (alors qu'il a déjà été set avant --- à mon avis un bug d'Android), ce qui coupe instantanément les sons qui sont en train de jouer (vu que `setType` appelle `stopAll`).
Encore pire (et ça c'est un bug d'Android), quand on sélectionne expressément le même item que déjà sélectionné dans la liste, le "onItemSelected" est même pas appelé ^^

Jimvy commented:
J'ai déjà un fix, je push sur cette branche ? 😁

Jimvy commented:
Ok, j'ai deux fix possibles :
- soit on passe "false" au `setSelection` de la ligne 32, ce qui désactive l'animation et, étrangement, désactive aussi l'appel au listener
- soit on met un if dans le listener, pour ne changer la valeur que si elle diffère de la valeur actuelle.

Jimvy committed:
Fix stop of sounds when opening ParameterFragment.

Now you need to select something different than the current mode to actually stop the sound playing.
Only one of the two fixes is actually needed; but as it's Android, I guess the current behaviour will change in the future, so let's be safe.

TGLuis commented:
Personnellement j'ai un autre truc étrange, le stopAll est même pas appelé sur mon gsm (même si on séléctionne un son différent) (ce qui était un bug mais qu'on a considéré comme une feature). D'après moi, si tu veux push le fix, 

1.  Tu enlèves le `stopAll` de `setType`
2. Tu ajoutes le false au `setSelection`

Et on devrait être bon :) Qu'est-ce que t'en penses?

Jimvy commented:
J'ai pourtant l'impression que le stopPlayer/stopAll est bien appelé chez moi quand j'ouvre le fragment...
Perso je serais plus en faveur de faire un if, si l'API de Google change, c'est ce qui a le moins de risque de disparaître.

Jimvy committed:
Changing the playing mode doesn't stop the sounds, to enable creative bugs 🙃
